test/parse_test.py

```diff
@@ -361,6 +361,126 @@ def test_inheritance_instantiation(self):
         # self.assertEqual(flat_tree.classes["M"].symbols["b.a.p"].type.name, "Integer")
         # self.assertEqual(flat_tree.classes["M"].symbols["b.a.p"].value.value, 1)
 
+    def test_extends_lookup_not_in_extended(self):
+        """
+        Check that we do not find 'model B' in the unnamed node 'A' in the
+        following model. We also check that the order of inheritance does not
+        matter for the error message.
+        """
+        txt = """
+        model A
+            model B
+                parameter Real x = 1.0;
+            end B;
+            parameter Real y = 2.0;
+        end A;
+
+        model C
+            extends {};
+            extends {};
+        end C;
+        """
+
+        for extends_1, extends_2 in [("A", "B"), ("B", "A")]:
+            ast_tree = parser.parse(txt.format(extends_1, extends_2))
+            instance_tree = tree.InstanceTree(ast_tree)
+
+            with self.assertRaisesRegex(
+                tree.ModelicaSemanticError, "Extends name B not found in scope C"
+            ):
+                instance = instance_tree.instantiate("C")  # noqa: F841
+
+    def test_error_extends_class_also_extended_name_simple(self):
+        """
+        Check that we are not allowed to inherit from `model B`, because a
+        symbol with the same name is inherited from `model A`. We also check
+        that the order of inheritance does not matter for the error message.
+        """
+        txt = """
+        model A
+            parameter Real B = 1.0;
+            parameter Real y = 2.0;
+        end A;
+
+        model B
+            parameter Real x = 1.0;
+        end B;
+
+        model C
+            extends {};
+            extends {};
+        end C;
+        """
+
+        for extends_1, extends_2 in [("A", "B")]:
+            ast_tree = parser.parse(txt.format(extends_1, extends_2))
+            instance_tree = tree.InstanceTree(ast_tree)
+
+            with self.assertRaisesRegex(
+                tree.ModelicaSemanticError,
+                "Cannot extend 'C' with 'B'; 'B' also exists in names inherited from 'A'",
+            ):
+                instance = instance_tree.instantiate("C")  # noqa: F841
+
+    def test_error_extends_class_also_extended_name_of_self(self):
+        """
+        Check that we are not allowed to inherit from `model A`, because a
+        symbol with the same name is inherited from `model A`.
+        """
+        txt = """
+        model A
+            parameter Real A = 1.0;
+            parameter Real y = 2.0;
+        end A;
+
+        model C
+            extends A;
+        end C;
+        """
+
+        ast_tree = parser.parse(txt)
+        instance_tree = tree.InstanceTree(ast_tree)
+
+        with self.assertRaisesRegex(
+            tree.ModelicaSemanticError,
+            "Cannot extend 'C' with 'A'; 'A' also exists in names inherited from 'A'",
+        ):
+            instance = instance_tree.instantiate("C")  # noqa: F841
```

Comment:
This does not look right to me, even though OpenModelica gives an error also: "Expected A to be a class, but found component instead.". The `extends A` should find `model A` and it stops there. `model A` has not already been inherited by `model C`, so it is OK. The `parameter Real A` is another level down in the hierarchy and does not clash. There is something in the spec about classes and symbols being in the same name space, but that is at any given level, not between levels in the hierarchy. Am I missing something here?

Reply:
BTW, I really like your nice, simple example here! Makes it easy to think about and debug (though I haven't done the latter).

Reply:
Thought about this one more and it actually does seem to adhere to the statement in the ModelicaCompliance example: "a name used in an extends-clause may not be inherited" exactly because symbols and classes are in the same namespace. But the example given in the ModelicaCompliance test is a little different because it is across *separate* extends clauses, not the same one. 

This is a tricky case. I wonder if this was discussed in the development of the spec. It's worth some more investigation, but I have spent quite a bit of time investigating and would rather move the code forward. If we think OpenModelica is a good representation of "correct", then this test is right. They do have a rather large test suite of open source libraries they test against, but the last time I looked (admittedly a while ago), they still fail on several. They are certainly lot better tested on real code than we are currently. 😅 

Maybe a good approach to just take what OpenModelica gives as "correct" for now to save time on spec-chasing, and circle back around when we get to the point of getting close to the OpenModelica level of testing.

Reply:
Yep, it's just one of the examples I thought of while implementing this. Checked against the comment again, and against OMC, and then added the test case.